### PR TITLE
feature: support globalnet in submariner-addon

### PR DESCRIPTION
**Pull request by SataQiu**
Fixes: #223

### CODE REVIEW

**Review by sridhargaddam — COMMENTED** [review]
Nice work SataQiu. Thanks for pushing the PR.

Some minor comments.

File: pkg/hub/submarinerbrokerinfo/brokerinfo.go [hunk 1]
```diff
@@ -176,6 +182,10 @@ func applySubmarinerConfig(
 		brokerInfo.SubmarinerRouteAgentImage = submarinerConfig.Spec.ImagePullSpecs.SubmarinerRouteAgentImagePullSpec
 	}
 
+	if submarinerConfig.Spec.ImagePullSpecs.SubmarinerGlobalNetImagePullSpec != "" {
+		brokerInfo.SubmarinerGlobalNetImage = submarinerConfig.Spec.ImagePullSpecs.SubmarinerGlobalNetImagePullSpec
+	}
```
[the remark this answers is not in the document]
Comment (SataQiu) [reply pkg/hub/submarinerbrokerinfo/brokerinfo.go]: Thanks for your detailed review @sridhargaddam 
All the comments have been addressed.

**Review by sridhargaddam — APPROVED** [review]
Good work @SataQiu. 

This PR addresses the following "Addon changes to support globalnet configuration while deploying managed clusters." Are you also planning to push PRs for automatically allocating globalCIDRs to managed clusters (just like how subctl join does) from the globalnet-cidr-range? 

EPIC: https://github.com/submariner-io/enhancements/issues/51

File: pkg/apis/submarinerconfig/v1alpha1/types.go [hunk 2]
```diff
@@ -73,12 +83,6 @@ type SubmarinerConfigSpec struct {
 	// GatewayConfig represents the gateways configuration of the Submariner.
 	// +optional
 	GatewayConfig `json:"gatewayConfig,omitempty"`
-
-	// LoadBalancerEnable enables or disables load balancer mode. When enabled, a LoadBalancer is created in the
-	// submariner-operator namespace (default false).
-	// +optional
-	// +kubebuilder:default=false
-	LoadBalancerEnable bool `json:"loadBalancerEnable"`
```
Comment (SataQiu) [inline pkg/apis/submarinerconfig/v1alpha1/types.go:81]: Move this close to `NATTEnable` to fix the `golangci-lint` field alignment error.

File: pkg/hub/submarinerbrokerinfo/brokerinfo.go [hunk 3]
```diff
@@ -70,9 +72,9 @@ type SubmarinerBrokerInfo struct {
 	CatalogStartingCSV        string
 	SubmarinerGatewayImage    string
 	SubmarinerRouteAgentImage string
+	SubmarinerGlobalnetImage  string
 	LighthouseAgentImage      string
 	LighthouseCoreDNSImage    string
-	LoadBalancerEnabled       bool
```
Comment (SataQiu) [inline pkg/hub/submarinerbrokerinfo/brokerinfo.go:75]: Move this close to `NATEnabled` to fix the `golangci-lint` field alignment error.

File: pkg/hub/submarinerbrokerinfo/brokerinfo.go [hunk 4]
```diff
@@ -168,21 +174,7 @@ func applySubmarinerConfig(
 		brokerInfo.CatalogStartingCSV = submarinerConfig.Spec.SubscriptionConfig.StartingCSV
 	}
 
-	if submarinerConfig.Spec.ImagePullSpecs.SubmarinerImagePullSpec != "" {
```
Comment (SataQiu) [inline pkg/hub/submarinerbrokerinfo/brokerinfo.go:171]: Extraction into subfunction to fix the `golangci-lint` gocyclo error.

**Comment by SataQiu** [comment]
> Good work @SataQiu.
> 
> This PR addresses the following "Addon changes to support globalnet configuration while deploying managed clusters." Are you also planning to push PRs for automatically allocating globalCIDRs to managed clusters (just like how subctl join does) from the globalnet-cidr-range?
> 
> EPIC: [submariner-io/enhancements#51](https://github.com/submariner-io/enhancements/issues/51)

@sridhargaddam That sounds just the feature we need. I will try to submit another PR soon.

**Comment by SataQiu** [comment]
Hi @sridhargaddam 

The `ci/prow/sonarcloud` CI always failed with `failed to create credentials: could not read source credential: secrets "acm-sonarcloud-token" not found`.  What should I do to fix this error?

**Comment by sridhargaddam** [comment]
> 
> @sridhargaddam That sounds just the feature we need. I will try to submit another PR soon.

Awesome. Can you share some details about the PR that you will be working on?

Just to give you some context about the behavior with subctl CLI:
When using "subctl" cli to deploy brokers/join-clusters with globalnet enabled, we have a provision to specify (or override) the default `--globalnet-cidr-range`. Also, while joining a cluster to the broker, subctl will automatically allocate a non-overlapping globalCIDR from the `--globalnet-cidr-range`  to each of the cluster. The current PR, AFAIU will allow users to manually specify the globalCIDR to each managed cluster but will not auto-allocate a globalCIDR from the `--globalnet-cidr-range` pool.

So, is your next PR about the auto-allocation of globalCIDR to each managed cluster from the `--globalnet-cidr-range` pool? 

On a related note, are you planning to work on changes to ACM GUI (https://github.com/open-cluster-management/console.git project) as well? Can you please let me know? Thank you @SataQiu

**Comment by SataQiu** [comment]
>So, is your next PR about the auto-allocation of globalCIDR to each managed cluster from the --globalnet-cidr-range pool?

Yes, I think the next step is to automatically assign a `globalCIDR` to the managed cluster.

**Comment by skitt** [comment]
/unassign @qiujian16 
/lgtm
/approve

The Sonar error is unrelated, it’s being looked into.

**Review by tpantelis — APPROVED** [review]

**Comment by ray-harris** [comment]
There was an issue with the sonar cloud token in the Prow vault. Fixed it and it should work now.

/test sonarcloud

**Pull request merged**